Let CMake handle NCCL detection instead of our handcrafted Python script.

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22819 Add sanity checks for NCCL detection.
* **#22818 Let CMake handle NCCL detection instead of our handcrafted Python script.**

- USE_NCCL, USE_SYSTEM_NCCL, USE_STATIC_NCCL are made proper CMake build
  options that depend on other options.
- Detections using folders such as lib64, lib64/<arch> are automatically
  handled in find_library in CMake.
- NCCL_ROOT_DIR in findNCCL.cmake is changed to NCCL_ROOT, which is the
  standard CMake variable that specifies the root dir of a library (and
  therefore supports a wide range of automatic detection). E.g., see
  https://cmake.org/cmake/help/latest/envvar/PackageName_ROOT.html

---

How does the current code subsume all detections in the deleted `nccl.py`?

- The dependency of `USE_NCCL` on the OS and `USE_CUDA` is handled as dependency options in `CMakeLists.txt`.

- The main NCCL detection happens in [FindNCCL.cmake](https://github.com/pytorch/pytorch/blob/8377d4b32c12206a0f9401e81a5e5796c8fc01a8/cmake/Modules/FindNCCL.cmake), which is called by [nccl.cmake](https://github.com/pytorch/pytorch/blob/8377d4b32c12206a0f9401e81a5e5796c8fc01a8/cmake/External/nccl.cmake). When `USE_SYSTEM_NCCL` is false, the previous Python code defer the detection to `find_package(NCCL)`. The change in `nccl.cmake` retains this.

- `USE_STATIC_NCCL` in the previous Python code simply changes the name of the detected library. This is done in `IF (USE_STATIC_NCCL)`.

- Now we only need to look at how the lines below line 20 in `nccl.cmake` are subsumed. These lines list paths to header and library directories that NCCL headers and libraries may reside in and try to search these directories for the key header and library files in turn. These are done by `find_path` for headers and `find_library` for the library files in `FindNCCL.cmake`.
  * The call of [find_path](https://cmake.org/cmake/help/v3.8/command/find_path.html) (Search for `NO_DEFAULT_PATH` in the link) by default searches for headers in `<prefix>/include` for each `<prefix>` in `CMAKE_PREFIX_PATH` and `CMAKE_SYSTEM_PREFIX_PATH`. Like the Python code, this commit sets `CMAKE_PREFIX_PATH` to search for `<prefix>` in `NCCL_ROOT_DIR` and home to CUDA.  `CMAKE_SYSTEM_PREFIX_PATH` includes the standard directories such as `/usr/local` and `/usr`. `NCCL_INCLUDE_DIR` is also specifically handled.

  * Similarly, the call of [find_library](https://cmake.org/cmake/help/v3.8/command/find_library.html) (Search for `NO_DEFAULT_PATH` in the link) by default searches for libraries in directories including `<prefix>/lib` for each `<prefix>` in `CMAKE_PREFIX_PATH` and `CMAKE_SYSTEM_PREFIX_PATH`. But it also handles the edge cases intended to be solved in the Python code more properly:
     - It only searches for `<prefix>/lib64` (and `<prefix>/lib32`) if it is appropriate on the system.
     - It only searches for `<prefix>/lib/<arch>` for the right `<arch>`, unlike the Python code searches for `lib/<arch>` in a generic way (e.g., the Python code searches for `/usr/lib/x86_64-linux-gnu` but in reality systems have `/usr/lib/x86_64-some-customized-name-linux-gnu`, see https://unix.stackexchange.com/a/226180/38242 ).

---

Regarding for relevant issues:


- #12063 and #2877: These are properly handled, as explained in the updated comment.
- #2941 does not changes NCCL detection specifically for Windows (it changed CUDA detection).
- b7e258f81ef61d19b884194cdbcd6c7089636d46 A versioned library detection is added, but the order is reversed: The unversioned library becomes preferred. This is because normally unversioned libraries are linked to versioned libraries and preferred by users, and local installation by users are often unversioned. Like the document of [find_library](https://cmake.org/cmake/help/v3.8/command/find_library.html) suggests:

> When using this to specify names with and without a version suffix, we recommend specifying the unversioned name first so that locally-built packages can be found before those provided by distributions.

Differential Revision: [D16281714](https://our.internmc.facebook.com/intern/diff/D16281714)

xuhdev committed:
Let CMake handle NCCL detection instead of our handcrafted Python script.

- USE_NCCL, USE_SYSTEM_NCCL, USE_STATIC_NCCL are made proper CMake build
  options that depend on other options.
- Detections using folders such as lib64, lib64/<arch> are automatically
  handled in find_library in CMake.
- NCCL_ROOT_DIR in findNCCL.cmake is changed to NCCL_ROOT, which is the
  standard CMake variable that specifies the root dir of a library (and
  therefore supports a wide range of automatic detection). E.g., see
  https://cmake.org/cmake/help/latest/envvar/PackageName_ROOT.html

lemairecarl commented:
Hi! I'm wondering if this PR will fix my issue. I'm having problems compiling with libtorch on a Linux installation where CUDA is not in `/usr/local/cuda`.
I found a workaround: I remove the references to that folder in `share/cmake/Caffe2/Caffe2Targets.cmake`. I use this command:
```
sed -i -e 's/\/usr\/local\/cuda\/lib64\/libculibos.a;dl;\/usr\/local\/cuda\/lib64\/libculibos.a;//g' share/cmake/Caffe2/Caffe2Targets.cmake
```
Do you think it will handle my case?

xuhdev commented:
> Hi! I'm wondering if this PR will fix my issue. I'm having problems compiling with libtorch on a Linux installation where CUDA is not in `/usr/local/cuda`.
> I found a workaround: I remove the references to that folder in `share/cmake/Caffe2/Caffe2Targets.cmake`. I use this command:
> 
> ```
> sed -i -e 's/\/usr\/local\/cuda\/lib64\/libculibos.a;dl;\/usr\/local\/cuda\/lib64\/libculibos.a;//g' share/cmake/Caffe2/Caffe2Targets.cmake
> ```
> 
> Do you think this PR handles my case?

This PR is about NCCL detection, not CUDA. You can try to set `CUDA=/path/to/your/cuda/home`. If it still doesn't work, feel free to start a [discussion](https://discuss.pytorch.org/) or open a new issue!

xuhdev committed:
Update on "Let CMake handle NCCL detection instead of our handcrafted Python script."

Let CMake handle NCCL detection instead of our handcrafted Python script.

- USE_NCCL, USE_SYSTEM_NCCL, USE_STATIC_NCCL are made proper CMake build
  options that depend on other options.
- Detections using folders such as lib64, lib64/<arch> are automatically
  handled in find_library in CMake.
- NCCL_ROOT_DIR in findNCCL.cmake is changed to NCCL_ROOT, which is the
  standard CMake variable that specifies the root dir of a library (and
  therefore supports a wide range of automatic detection). E.g., see
  https://cmake.org/cmake/help/latest/envvar/PackageName_ROOT.html

gh-metadata: pytorch pytorch 22818 gh/xuhdev/13/head

xuhdev committed:
Update on "Let CMake handle NCCL detection instead of our handcrafted Python script."

Let CMake handle NCCL detection instead of our handcrafted Python script.

- USE_NCCL, USE_SYSTEM_NCCL, USE_STATIC_NCCL are made proper CMake build
  options that depend on other options.
- Detections using folders such as lib64, lib64/<arch> are automatically
  handled in find_library in CMake.
- NCCL_ROOT_DIR in findNCCL.cmake is changed to NCCL_ROOT, which is the
  standard CMake variable that specifies the root dir of a library (and
  therefore supports a wide range of automatic detection). E.g., see
  https://cmake.org/cmake/help/latest/envvar/PackageName_ROOT.html

gh-metadata: pytorch pytorch 22818 gh/xuhdev/13/head

ezyang commented:
After this PR the environment variable `NCCL_ROOT_DIR` is no longer respected. It should be, especially because gloo has its own copy of Findnccl and will respect the envvar. (Arguably, we should make sure that we feed gloo cmake our own detection instead of using its settings.)

ezyang commented on cmake/External/nccl.cmake:
```diff
@@ -2,22 +2,12 @@ if (NOT __NCCL_INCLUDED)
   set(__NCCL_INCLUDED TRUE)
 
   if (USE_SYSTEM_NCCL)
-    # if we have explicit paths passed from setup.py, use those
```
Good!

ezyang commented on cmake/Modules/FindNCCL.cmake:
```diff
@@ -14,32 +14,34 @@
+set(NCCL_LIB_DIR $ENV{NCCL_LIB_DIR} CACHE PATH "Folder contains NVIDIA NCCL libraries")
+set(NCCL_VERSION $ENV{NCCL_VERSION} CACHE STRING "Version of NCCL to build with")
+
+list(APPEND NCCL_ROOT ${NCCL_ROOT_DIR} ${CUDA_TOOLKIT_ROOT_DIR})
```
I think you probably meant `$ENV{NCCL_ROOT_DIR}` here? `NCCL_ROOT_DIR` is never set otherwise.

xuhdev replied:
Yes, nice catch. I changed this to ENV and added a deprecation message. It used to be set in cmake.py but somehow I lost it during rebase.

ezyang commented on cmake/Modules/FindNCCL.cmake:
```diff
@@ -14,32 +14,34 @@
-  ${NCCL_ROOT_DIR}
-  ${NCCL_ROOT_DIR}/include
-  ${CUDA_TOOLKIT_ROOT_DIR}/include)
+  HINTS ${NCCL_INCLUDE_DIR})
```
Good.

ezyang commented on cmake/Modules/FindNCCL.cmake:
```diff
@@ -14,32 +14,34 @@
   SET(NCCL_LIBNAME "nccl")
-ENDIF()
+  if (NCCL_VERSION)  # Prefer the versioned library if a specific NCCL version is specified
+    set(CMAKE_FIND_LIBRARY_SUFFIXES ".so.${NCCL_VERSION}" ${CMAKE_FIND_LIBRARY_SUFFIXES})
```
I'm confused about the placement of this line. It occurs inside the `USE_STATIC_NCCL` block. Doesn't that mean that it will never actually usefully trigger (because we'll be linking against static nccl?)

xuhdev replied:
I think this is under the `else`-clause? This line will be triggered only if `USE_STATIC_NCCL` is false.

ezyang commented on cmake/Modules/FindNCCL.cmake:
```diff
@@ -14,32 +14,34 @@
-  ${NCCL_ROOT_DIR}/lib/x86_64-linux-gnu
-  ${NCCL_ROOT_DIR}/lib64
-  ${CUDA_TOOLKIT_ROOT_DIR}/lib64)
+  HINTS ${NCCL_LIB_DIR})
```
I'm actually a little confused by what the manual says here; I think in practice this means that we will stop looking in `lib64` location because cmake will only look in that directory if `FIND_LIBRARY_USE_LIB64_PATHS` is set (and I don't think it is set on ordinary platforms). I don't think this matters though, because I checked some of my system nccl installs and they appear to be in non-lib64 directory.

xuhdev replied:
I think FIND_LIBRARY_USE_LIB64_PATHS is set when it is proper (such as a multiarch system) to do so. From the document of find_library:

> If the FIND_LIBRARY_USE_LIB64_PATHS global property is set all search paths will be tested as normal, with 64/ appended, and with all matches of lib/ replaced with lib64/. **This property is automatically set for the platforms that are known to need it** if at least one of the languages supported by the project() command is enabled.

ezyang reviewed (changes requested):
This looks very good but I am concerned about the loss of `NCCL_ROOT_DIR` envvar support. (The dynamic library version pinning is also probably broken, but that error seems pretty benign.)

xuhdev committed:
Update on "Let CMake handle NCCL detection instead of our handcrafted Python script."

Let CMake handle NCCL detection instead of our handcrafted Python script.

- USE_NCCL, USE_SYSTEM_NCCL, USE_STATIC_NCCL are made proper CMake build
  options that depend on other options.
- Detections using folders such as lib64, lib64/<arch> are automatically
  handled in find_library in CMake.
- NCCL_ROOT_DIR in findNCCL.cmake is changed to NCCL_ROOT, which is the
  standard CMake variable that specifies the root dir of a library (and
  therefore supports a wide range of automatic detection). E.g., see
  https://cmake.org/cmake/help/latest/envvar/PackageName_ROOT.html

gh-metadata: pytorch pytorch 22818 gh/xuhdev/13/head

xuhdev committed:
Update on "Let CMake handle NCCL detection instead of our handcrafted Python script."

Let CMake handle NCCL detection instead of our handcrafted Python script.

- USE_NCCL, USE_SYSTEM_NCCL, USE_STATIC_NCCL are made proper CMake build
  options that depend on other options.
- Detections using folders such as lib64, lib64/<arch> are automatically
  handled in find_library in CMake.
- NCCL_ROOT_DIR in findNCCL.cmake is changed to NCCL_ROOT, which is the
  standard CMake variable that specifies the root dir of a library (and
  therefore supports a wide range of automatic detection). E.g., see
  https://cmake.org/cmake/help/latest/envvar/PackageName_ROOT.html

gh-metadata: pytorch pytorch 22818 gh/xuhdev/13/head

xuhdev committed:
Update on "Let CMake handle NCCL detection instead of our handcrafted Python script."

Let CMake handle NCCL detection instead of our handcrafted Python script.

- USE_NCCL, USE_SYSTEM_NCCL, USE_STATIC_NCCL are made proper CMake build
  options that depend on other options.
- Detections using folders such as lib64, lib64/<arch> are automatically
  handled in find_library in CMake.
- NCCL_ROOT_DIR in findNCCL.cmake is changed to NCCL_ROOT, which is the
  standard CMake variable that specifies the root dir of a library (and
  therefore supports a wide range of automatic detection). E.g., see
  https://cmake.org/cmake/help/latest/envvar/PackageName_ROOT.html

gh-metadata: pytorch pytorch 22818 gh/xuhdev/13/head

xuhdev committed:
Update on "Let CMake handle NCCL detection instead of our handcrafted Python script."

Let CMake handle NCCL detection instead of our handcrafted Python script.

- USE_NCCL, USE_SYSTEM_NCCL, USE_STATIC_NCCL are made proper CMake build
  options that depend on other options.
- Detections using folders such as lib64, lib64/<arch> are automatically
  handled in find_library in CMake.
- NCCL_ROOT_DIR in findNCCL.cmake is changed to NCCL_ROOT, which is the
  standard CMake variable that specifies the root dir of a library (and
  therefore supports a wide range of automatic detection). E.g., see
  https://cmake.org/cmake/help/latest/envvar/PackageName_ROOT.html

gh-metadata: pytorch pytorch 22818 gh/xuhdev/13/head

xuhdev committed:
Update on "Let CMake handle NCCL detection instead of our handcrafted Python script."

Let CMake handle NCCL detection instead of our handcrafted Python script.

- USE_NCCL, USE_SYSTEM_NCCL, USE_STATIC_NCCL are made proper CMake build
  options that depend on other options.
- Detections using folders such as lib64, lib64/<arch> are automatically
  handled in find_library in CMake.
- NCCL_ROOT_DIR in findNCCL.cmake is changed to NCCL_ROOT, which is the
  standard CMake variable that specifies the root dir of a library (and
  therefore supports a wide range of automatic detection). E.g., see
  https://cmake.org/cmake/help/latest/envvar/PackageName_ROOT.html

gh-metadata: pytorch pytorch 22818 gh/xuhdev/13/head

xuhdev commented:
> After this PR the environment variable `NCCL_ROOT_DIR` is no longer respected. It should be, especially because gloo has its own copy of Findnccl and will respect the envvar. (Arguably, we should make sure that we feed gloo cmake our own detection instead of using its settings.)

Do you also wanna use this findNCCL once review is finished? This copy of findNCCL is actually pretty generic.

ezyang commented:
> Do you also wanna use this findNCCL once review is finished? This copy of findNCCL is actually pretty generic.

I think the gloo bits live in another repository, so we can't easily replace it.

zou3519 commented:
@ezyang could you land this when you get a chance?

zhangguanheng66 commented:
@ezyang Same request as Richard. Thanks.

xuhdev commented:
I reopened this is #22930 . For some reason this actually has already been landed but was reverted later. The stack here is actually meaningless now.

ezyang commented:
Thanks, I'll land that one.
